### PR TITLE
Refactors leftover from PR 5356

### PR DESCRIPTION
**Pull request by wrzontek**
## Describe your changes
see comments in https://github.com/TouK/nussknacker/pull/5356

## Checklist before merge
- [ ] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [ ] Code is cleaned from temporary changes and commented out lines
- [ ] Parts of the code that are not easy to understand are documented in the code
- [ ] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [ ] Verify that PR will be squashed during merge

### CODE REVIEW

File: interpreter/src/main/scala/pl/touk/nussknacker/engine/testing/ModelDefinitionBuilder.scala [hunk 1]
```diff
@@ -54,19 +58,29 @@ final case class ModelDefinitionBuilder(
       name: String,
       returnType: Option[TypingResult],
       componentSpecificData: CustomComponentSpecificData,
-      params: Parameter*
+      componentId: Option[ComponentId],
```
Comment (arkadius) [inline interpreter/src/main/scala/pl/touk/nussknacker/engine/testing/ModelDefinitionBuilder.scala:61]: Please move componentId on the and, add the default `None` and remove all duplications of `componentId = None`. Thanks to that we won't spread this edge case through the all places where it is not important.

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/component/ComponentsUsageHelperTest.scala [hunk 2]
```diff
@@ -131,11 +139,49 @@ class ComponentsUsageHelperTest extends AnyFunSuite with Matchers with TableDriv
 
   private val processDetailsWithFragment = displayableToProcess(TestProcessUtil.toDisplayable(processWithFragment))
 
-  private val processingTypeAndComponentInfoToComponentId: (ProcessingType, ComponentInfo) => ComponentId = {
-    case (_, ComponentInfo(ComponentType.CustomComponent, `otherExistingStreamTransformer`)) =>
-      ComponentId(overriddenOtherExistingStreamTransformer)
-    case (processingType, info) =>
-      ComponentId.default(processingType, info)
+  private def nonFragmentComponents(componentInfoToId: ComponentInfo => ComponentId) = {
+    val modelDefinition = ModelDefinitionBuilder
```
Comment (arkadius) [inline designer/server/src/test/scala/pl/touk/nussknacker/ui/component/ComponentsUsageHelperTest.scala:143]: From what place did you duplicate this code? I won't believe that you wrote it from scratch :) Can you deduplicate this code?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentService.scala [hunk 3]
```diff
@@ -65,29 +66,12 @@ class DefaultComponentService(
 
   import cats.syntax.traverse._
 
-  private def processingTypeAndComponentInfoToComponentId(
+  private def processingTypeToNonFragmentComponents(
```
Comment (arkadius) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentService.scala:69]: 1. Please write a comment that we collect all components but fragments because fragments can't have ComponentId overridden, so we can use the default one without fetching them. 
2. Please move this method below the last invocation of it
3. After 2., please also move  `definedComponents` below it's last invocation.
Thanks to 2. and 3., we reduce the number of jumps thought the file during further code analysis

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentsUsageHelper.scala [hunk 4]
```diff
@@ -12,21 +13,24 @@ object ComponentsUsageHelper {
...
   ): Map[ComponentId, List[(ScenarioWithDetailsEntity[_], List[NodeUsageData])]] = {
     def flattenUsages(processesDetails: List[ScenarioWithDetailsEntity[ScenarioComponentsUsages]]) = for {
       processDetails    <- processesDetails
       componentInfoNode <- processDetails.json.value.toList
       (componentInfo, nodeIds) = componentInfoNode
-      componentId = processingTypeAndComponentInfoToComponentId(processDetails.processingType, componentInfo)
+      componentId = processingTypeToNonFragmentComponents(processDetails.processingType)
+        .get(componentInfo)
+        .map(_.componentIdUnsafe)
+        .getOrElse(ComponentId.default(processDetails.processingType, componentInfo))
```
Comment (arkadius) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentsUsageHelper.scala:33]: Please write a comment that we pick the default one instead of skipping this usage because for fragments, there won't be an id in the `processingTypeToNonFragmentComponents`

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentsUsageHelper.scala [hunk 5]
```diff
@@ -12,21 +13,24 @@ object ComponentsUsageHelper {
 
   def computeComponentsUsageCount(
       processesDetails: List[ScenarioWithDetailsEntity[ScenarioComponentsUsages]],
-      processingTypeAndComponentInfoToComponentId: (ProcessingType, ComponentInfo) => ComponentId,
+      processingTypeToNonFragmentComponents: Map[ProcessingType, Map[ComponentInfo, ComponentStaticDefinition]]
```
Comment (arkadius) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentsUsageHelper.scala:16]: I know that I proposed similar approach during the previous review, but I see know that this signature looks ugly. can you flatten this map to `Map[(ProcessingType, ComponentInfo), ComponentId]?`

File: interpreter/src/main/scala/pl/touk/nussknacker/engine/testing/ModelDefinitionBuilder.scala (hunk 1 above)
Comment (wrzontek) [reply interpreter/src/main/scala/pl/touk/nussknacker/engine/testing/ModelDefinitionBuilder.scala:61]: I've tried to add a default case, but scala isn't too happy with it, even though it should be easy for it to differentiate between type `Parameter` and type `Option[ComponentId]`

- *-parameter must come last
- Parameter section with *-parameter is not allowed to have default arguments

Comment (wrzontek) [reply interpreter/src/main/scala/pl/touk/nussknacker/engine/testing/ModelDefinitionBuilder.scala:61]: ok i have an idea:

```
  def withCustom(
      name: String,
      returnType: Option[TypingResult],
      componentSpecificData: CustomComponentSpecificData,
      componentId: Option[ComponentId] = None,
  ): ModelDefinitionBuilder =
    withCustom(name, returnType, componentSpecificData, componentId, Seq.empty: _*)

  def withCustom(
      name: String,
      returnType: Option[TypingResult],
      componentSpecificData: CustomComponentSpecificData,
      componentId: Option[ComponentId],
      params: Parameter*
  ): ModelDefinitionBuilder =
    wrapWithNotDisabledCustomComponentDefinition(
      params.toList,
      returnType,
      componentSpecificData
    ).map(withComponent(name, _, componentId)).getOrElse(this)

```

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/component/ComponentsUsageHelperTest.scala (hunk 2 above)
Comment (wrzontek) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/component/ComponentsUsageHelperTest.scala:143]: i'ts not really duplication, i've used ModelDefinitionBuilder analogously to ProcessTestData/ProcessValidatorSpec but included only the components needed by the test

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentsUsageHelper.scala (hunk 5 above)
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentsUsageHelper.scala]: I like it, let it stay

**Review by arkadius — APPROVED** [review]
Good job, thanks
